Create and import hosts

stejskalleos opened the pull request:
Note: Without the Cloud-init scenario

stejskalleos committed:
Create and import hosts

adamruzicka commented:
Creating a VM with multiple disks doesn't seem to work:
```
2022-06-22T14:24:54 [W|app|49654396] »Failed to create a compute foreman-google (us-central1-a-Google) instance rose-tittl.addie-dulberg.example.com: ERF50-2985 [Foreman::WrappedException]: »Could not create Google resource disks« ([Google::Cloud::InvalidArgumentError]: An error 
has occurred when making a REST request: Invalid value for field 'resource.sourceImage': ''. The URL must have at least 1 character(s))                                                                                                                                                 
 49654396 |  «                                                                                                                                                                                                                                                                          
2022-06-22T14:24:54 [I|app|49654396] Backtrace for '»Failed to create a compute foreman-google (us-central1-a-Google) instance rose-tittl.addie-dulberg.example.com: ERF50-2985 [Foreman::WrappedException]: »Could not create Google resource disks« ([Google::Cloud::InvalidArgumentEr
ror]: An error has occurred when making a REST request: Invalid value for field 'resource.sourceImage': ''. The URL must have at least 1 character(s))                                                                                                                                  
 49654396 |  «' error (Foreman::WrappedException): ERF50-2985 [Foreman::WrappedException]: »Could not create Google resource disks« ([Google::Cloud::InvalidArgumentError]: An error has occurred when making a REST request: Invalid value for field 'resource.sourceImage': ''. The UR
L must have at least 1 character(s))                                                                                                                                                                                                                                                    
 49654396 | /home/aruzicka/vcs/foreman/foreman-google/app/lib/foreman_google/google_compute_adapter.rb:131:in `rescue in insert'                                                                                                                                                        
 49654396 | /home/aruzicka/vcs/foreman/foreman-google/app/lib/foreman_google/google_compute_adapter.rb:128:in `insert'                                                                                                                                                                  
 49654396 | /home/aruzicka/vcs/foreman/foreman-google/app/lib/foreman_google/google_compute_adapter.rb:71:in `insert_disk'                                                                                                                                                              
 49654396 | /home/aruzicka/vcs/foreman/foreman-google/app/models/foreman_google/google_compute.rb:70:in `block in create_volumes'                                                                                                                                                       
 49654396 | /home/aruzicka/vcs/foreman/foreman-google/app/models/foreman_google/google_compute.rb:67:in `each'                                                                                                                                                                          
 49654396 | /home/aruzicka/vcs/foreman/foreman-google/app/models/foreman_google/google_compute.rb:67:in `create_volumes'                                                                                                                                                                 49654396 | /home/aruzicka/vcs/foreman/foreman-google/app/models/foreman_google/gce.rb:93:in `create_vm'
 ```

adamruzicka commented:
VM removal seems to be broken now, VMs created from foreman cannot be removed from it again (even if they are already removed from the cloud)
```
2022-06-22T14:30:17 [D|app|fa2260da] Processing task '»Removing compute instance brian-matusz.addie-dulberg.example.com«' from 'Host::Managed Main'    
2022-06-22T14:30:17 [I|app|fa2260da] Removing Compute instance for brian-matusz.addie-dulberg.example.com    
2022-06-22T14:30:17 [W|app|fa2260da] Rolling back due to a problem: [#<Orchestration::Task:0x00000000134dca98 @name="»Removing compute instance brian-matusz.addie-dulberg.example.com«", @id="»Removing compute instance brian-matusz.addie-dulberg.example.com«", @status="failed", @p
riority=100, @action=[#<Host::Managed id: 107, name: "brian-matusz.addie-dulberg.example.com", last_compile: nil, last_report: nil, updated_at: "2022-06-22 14:16:25.122669000 +0200", created_at: "2022-06-22 14:16:25.122669000 +0200", root_pass: [FILTERED], architecture_id: 1, ope
ratingsystem_id: 4, ptable_id: nil, medium_id: nil, build: true, comment: "", disk: "", installed_at: nil, model_id: nil, hostgroup_id: 2, owner_id: 4, owner_type: "User", enabled: true, puppet_ca_proxy_id: nil, managed: true, use_image: nil, image_file: nil, uuid: "brian-matusz-
addie-dulberg-example-com", compute_resource_id: 1, puppet_proxy_id: nil, certname: nil, image_id: 2, organization_id: 1, location_id: 2, type: "Host::Managed", otp: nil, realm_id: nil, compute_profile_id: nil, provision_method: "image", grub_pass: nil, global_status: 0, lookup_v
alue_matcher: [FILTERED], pxe_loader: "", initiated_at: nil, build_errors: nil>, :delCompute], @created=1655901017.703678, @timestamp=2022-06-22 12:30:17.745323684 UTC>]
2022-06-22T14:30:17 [I|app|fa2260da] Processed 1 tasks from queue 'Host::Managed Main', completed 0/1                                                                                                                                                                                   
2022-06-22T14:30:17 [E|app|fa2260da] Task '»Removing compute instance brian-matusz.addie-dulberg.example.com«' *failed*                                                                                                                                                                 
2022-06-22T14:30:17 [D|sql|fa2260da]   TRANSACTION (0.2ms)  ROLLBACK                                                                                                                                                                                                                    
2022-06-22T14:30:17 [D|sql|fa2260da]   ↳ app/controllers/api/v2/hosts_controller.rb:172:in `destroy'                                                                                                                                                                                    
2022-06-22T14:30:17 [W|app|fa2260da] Action failed                                                                                                                                                                                                                                      
2022-06-22T14:30:17 [I|app|fa2260da] Backtrace for 'Action failed' error (RuntimeError): resource have no errors                                                                                                                                                                        
 fa2260da | /home/aruzicka/vcs/foreman/foreman/app/controllers/api/base_controller.rb:155:in `process_resource_error'         
 fa2260da | /home/aruzicka/vcs/foreman/foreman/app/controllers/api/base_controller.rb:175:in `process_response'           
 fa2260da | /home/aruzicka/vcs/foreman/foreman/app/controllers/api/v2/hosts_controller.rb:172:in `destroy'
```

adamruzicka commented on app/models/foreman_google/google_compute.rb:
```diff
@@ -167,25 +172,28 @@ def load_image(image_id)
-      return if user_data.blank?
-      { items: [{ key: 'user-data', value: user_data }] }
+    def construct_metadata(args)
+      ssh_keys = { key: 'ssh-keys', value: "#{args[:username]}:#{args[:public_key]}" }
```
Not sure what to think about this. When the vm is created, the ssh post-provisioning fails to connect. Yet when I access the host from gce console, the key seems to be present in root's authorized_keys file

stejskalleos replied:
> When the vm is created, the ssh post-provisioning fails to connect

Really? It worked for me well, it takes some time until instance is reachable via SSH but it works on my env

adamruzicka replied:
The instance was up, but the key it tried to use was not accepted

stejskalleos replied:
Hmm interesting, can you send me vm data, os and so on so I could reproduce it?

adamruzicka replied:
CentOS7, image centos-7-v20220621, without userdata, deployed into us-central-1 as e2-micro with an ephemeral public ip.

```
D, [2022-06-22T16:43:18.306775 #261676] DEBUG -- net.ssh.authentication.methods.publickey[f8a5c]: trying publickey (d3:da:11:fc:06:6b:84:7c:e3:a8:1b:f8:f4:53:c5:8e)                                                                                                                    
2022-06-22T16:43:18 [D|app|14b6dc5e] trying publickey (d3:da:11:fc:06:6b:84:7c:e3:a8:1b:f8:f4:53:c5:8e)                                                                                                                                                                                 
D, [2022-06-22T16:43:18.306986 #261676] DEBUG -- socket[f89e4]: queueing packet nr 5 type 50 len 348                                        
2022-06-22T16:43:18 [D|app|14b6dc5e] queueing packet nr 5 type 50 len 348                                                                                                                                                                                                               
D, [2022-06-22T16:43:18.307145 #261676] DEBUG -- socket[f89e4]: sent 372 bytes                                                                                                                                                                                                          
2022-06-22T16:43:18 [D|app|14b6dc5e] sent 372 bytes                                                                                                                                                                                                                                     
D, [2022-06-22T16:43:18.485333 #261676] DEBUG -- socket[f89e4]: read 324 bytes                                                                                                                                                                                                          
2022-06-22T16:43:18 [D|app|14b6dc5e] read 324 bytes                                                                                                                                                                                                                                     
D, [2022-06-22T16:43:18.485533 #261676] DEBUG -- socket[f89e4]: received packet nr 5 type 60 len 300                                                                                                                                                                                    
2022-06-22T16:43:18 [D|app|14b6dc5e] received packet nr 5 type 60 len 300                                                                                                                                                                                                               
D, [2022-06-22T16:43:18.486967 #261676] DEBUG -- socket[f89e4]: queueing packet nr 6 type 50 len 620                                                                                                                                                                                    
2022-06-22T16:43:18 [D|app|14b6dc5e] queueing packet nr 6 type 50 len 620                                                                                                                                                                                                               
D, [2022-06-22T16:43:18.487115 #261676] DEBUG -- socket[f89e4]: sent 644 bytes                                                                                                                                                                                                          
2022-06-22T16:43:18 [D|app|14b6dc5e] sent 644 bytes                                                                                                                                                                                                                                     
D, [2022-06-22T16:43:18.618956 #261676] DEBUG -- socket[f89e4]: read 84 bytes                                                                                                                                                                                                           
2022-06-22T16:43:18 [D|app|14b6dc5e] read 84 bytes                                                                                                                                                                                                                                      
D, [2022-06-22T16:43:18.619181 #261676] DEBUG -- socket[f89e4]: received packet nr 6 type 51 len 60                                                                                                                                                                                     
2022-06-22T16:43:18 [D|app|14b6dc5e] received packet nr 6 type 51 len 60                                                                                                                                                                                                                
D, [2022-06-22T16:43:18.619336 #261676] DEBUG -- net.ssh.authentication.session[f8a20]: allowed methods: publickey,gssapi-keyex,gssapi-with-mic                                                                                                                                         
2022-06-22T16:43:18 [D|app|14b6dc5e] allowed methods: publickey,gssapi-keyex,gssapi-with-mic                                                                                                                                                                                            
D, [2022-06-22T16:43:18.619447 #261676] DEBUG -- net.ssh.authentication.methods.publickey[f8a5c]: publickey failed (d3:da:11:fc:06:6b:84:7c:e3:a8:1b:f8:f4:53:c5:8e)                                                                                                                    
2022-06-22T16:43:18 [D|app|14b6dc5e] publickey failed (d3:da:11:fc:06:6b:84:7c:e3:a8:1b:f8:f4:53:c5:8e)                                                                                                                                                                                 
E, [2022-06-22T16:43:18.619552 #261676] ERROR -- net.ssh.authentication.session[f8a20]: all authorization methods failed (tried publickey)                                                                                                                                              
2022-06-22T16:43:18 [E|app|14b6dc5e] all authorization methods failed (tried publickey)                                                                                                                                                                                                 
2022-06-22T16:43:18 [D|app|14b6dc5e] Auth failed for root at 35.202.149.249, retrying
```

adamruzicka commented:
Creating VMs from some images (hit with debian-9-stretch-v20220621) sometimes fails with
```
2022-06-22T13:42:51 [D|app|26c3d694] Processing task '»Set up compute instance joan-dargis.addie-dulberg.example.com«' from 'Host::Managed Main'
2022-06-22T13:42:51 [I|app|26c3d694] Adding Compute instance for joan-dargis.addie-dulberg.example.com
2022-06-22T13:42:51 [D|sql|26c3d694]   Image Load (0.4ms)  SELECT "images".* FROM "images" WHERE "images"."compute_resource_id" = $1 AND "images"."uuid" = $2 LIMIT $3  [["compute_resource_id", 1], ["uuid", "7619786046600874598"], ["LIMIT", 1]]
2022-06-22T13:42:51 [D|sql|26c3d694]   ↳ /home/aruzicka/vcs/foreman/foreman-google/app/models/foreman_google/gce.rb:90:in `create_vm'
2022-06-22T13:42:51 [D|sql|26c3d694]   KeyPair Load (4.7ms)  SELECT "key_pairs".* FROM "key_pairs" WHERE "key_pairs"."compute_resource_id" = $1 LIMIT $2  [["compute_resource_id", 1], ["LIMIT", 1]]
2022-06-22T13:42:51 [D|sql|26c3d694]   ↳ /home/aruzicka/vcs/foreman/foreman-google/app/models/foreman_google/gce.rb:90:in `create_vm'
2022-06-22T13:43:03 [W|app|26c3d694] »Failed to create a compute foreman-google (us-central1-a-Google) instance joan-dargis.addie-dulberg.example.com: ERF50-1985 [Foreman::WrappedException]: »Cannot insert instance!« ([Google::Cloud::InvalidArgumentError]: An error has occurred w
hen making a REST request: Instances with guest accelerators do not support live migration.)
 26c3d694 |  «
2022-06-22T13:43:03 [I|app|26c3d694] Backtrace for '»Failed to create a compute foreman-google (us-central1-a-Google) instance joan-dargis.addie-dulberg.example.com: ERF50-1985 [Foreman::WrappedException]: »Cannot insert instance!« ([Google::Cloud::InvalidArgumentError]: An error
 has occurred when making a REST request: Instances with guest accelerators do not support live migration.)
 26c3d694 |  «' error (Foreman::WrappedException): ERF50-1985 [Foreman::WrappedException]: »Cannot insert instance!« ([Google::Cloud::InvalidArgumentError]: An error has occurred when making a REST request: Instances with guest accelerators do not support live migration.)
 26c3d694 | /home/aruzicka/vcs/foreman/foreman-google/app/models/foreman_google/gce.rb:100:in `rescue in create_vm'
 26c3d694 | /home/aruzicka/vcs/foreman/foreman-google/app/models/foreman_google/gce.rb:89:in `create_vm'
 26c3d694 | /home/aruzicka/vcs/foreman/foreman/app/models/concerns/orchestration/compute.rb:107:in `setCompute'
 26c3d694 | /home/aruzicka/vcs/foreman/foreman/app/models/concerns/orchestration.rb:227:in `execute'
 26c3d694 | /home/aruzicka/vcs/foreman/foreman/app/models/concerns/orchestration.rb:152:in `block in process'
 26c3d694 | /home/aruzicka/vcs/foreman/foreman/app/models/concerns/orchestration.rb:144:in `each'
 26c3d694 | /home/aruzicka/vcs/foreman/foreman/app/models/concerns/orchestration.rb:144:in `process'
 26c3d694 | /home/aruzicka/vcs/foreman/foreman/app/models/concerns/orchestration.rb:44:in `around_save_orchestration'
```

adamruzicka commented:
Associating an vm doesn't seem to work
```
2022-06-23T09:46:19 [D|sql|f767f981]   ↳ app/controllers/compute_resources_vms_controller.rb:50:in `associate'
2022-06-23T09:46:19 [I|app|f767f981] Completed 500 Internal Server Error in 358ms (ActiveRecord: 5.3ms | Allocations: 54692)
 f767f981 | 
 f767f981 | 
2022-06-23T09:46:19 [F|app|f767f981]   
 f767f981 | NoMethodError (undefined method `public_ip_address' for #<ForemanGoogle::GoogleCompute:0x0000000014600d10>
 f767f981 | Did you mean?  public_send):
 f767f981 |   
 f767f981 | /home/aruzicka/vcs/foreman/foreman-google/app/models/foreman_google/gce.rb:129:in `associated_host'
 f767f981 | app/controllers/compute_resources_vms_controller.rb:54:in `associate'
 f767f981 | app/controllers/concerns/foreman/controller/timezone.rb:10:in `set_timezone'
 f767f981 | app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 f767f981 | app/controllers/concerns/foreman/controller/topbar_sweeper.rb:12:in `set_topbar_sweeper_controller'
 f767f981 | lib/foreman/middleware/libvirt_connection_cleaner.rb:9:in `call'
 f767f981 | lib/foreman/middleware/telemetry.rb:10:in `call'
 f767f981 | lib/foreman/middleware/logging_context_session.rb:22:in `call'
 f767f981 | lib/foreman/middleware/logging_context_request.rb:11:in `call'
 ```

stejskalleos committed:
C/I - issue with multiple disks and missing boot image

stejskalleos committed:
C/I - Add info about ssh for root user in image

stejskalleos committed:
C/I - Fix associate host feature

stejskalleos committed:
C/I - make rubocop happy

stejskalleos commented:
@adamruzicka 

* Disk issues - fixed with tests
* Removing hosts - works for me, tried with enabled / disabled `destroy_vm_on_host_delete` setting
* SSH issue - it's because you use `root` user, for which GCE doesn't allow SSH access. I added new `app/views/images/form/_gce.html.erb` template where default user is not `root` anymore and updated the form inline help
* Debian issue - it does happen only in `us-central1-a` zone, in `europe-central` it's fine :facepalm:, so my guess would be something with [GPU quotas per zone](https://cloud.google.com/compute/docs/gpus/create-vm-with-gpus#host-maintenance)
* Balanced vs Standard disks - I'm looking into it, will fix in different PR
* Associate VM - fixed

adamruzicka commented:
Removal mystery solved, I had too old base on which I was applying patches from here.

adamruzicka commented:
Thank you @stejskalleos !
